### PR TITLE
[Test Only] Eltwise Unary bfloat16 coverage tests - category 4 ops

### LINKED ISSUE #56823: #53613: Category 4 — Exhaustive BF16 coverage for unary ops with a float/scalar parameter
## Parent Issue
Closes / contributes to #53613 (Eltwise unit tests restructuring — standardised Eltwise tests)

## PR
#56947

## Summary

Adds exhaustive bfloat16 accuracy coverage for the 11 **Category-4 unary ops** that each
take a single tensor input plus one Python float/scalar parameter: `ttnn.elu`,
`ttnn.heaviside`, `ttnn.leaky_relu`, `ttnn.relu_max`, `ttnn.relu_min`, `ttnn.rpow`,
`ttnn.celu`, `ttnn.softcap`, `ttnn.fill`, `ttnn.hardshrink`, `ttnn.softshrink`.

Unlike Category 5's dim-split gated-linear-units, the input tensor is where "exhaustive
bfloat16 coverage" applies here: every op sweeps all 65,024 normal bfloat16 bit-patterns
through the tensor argument (one device dispatch per (op, param) pair). The scalar
parameter is not itself tensor-shaped, so it is parametrized over a handful of
representative values per op (default, fractional, integer, sign-varying, plus
non-positive alpha for elu) instead — except `fill`, whose output does not depend on
the input tensor at all; there `fill_value` is instead parametrized across representative
bfloat16 value classes (zero, signed, tiny/max magnitude, non-representable), **not**
an exhaustive sweep of all 65,536 possible `fill_value` bit-patterns (that would require
one device dispatch per fill_value, which isn't practical for CI).

| Op | Parameter | Accuracy criterion |
|---|---|---|
| `ttnn.heaviside` | `value` | exact (bit-for-bit), excluding `-0.0` (hardware sign-bit quirk) |
| `ttnn.relu_max` | `upper_limit` | exact (bit-for-bit) |
| `ttnn.relu_min` | `lower_limit` | exact (bit-for-bit) |
| `ttnn.hardshrink` | `lambd` (default 0.5) | exact (bit-for-bit); `lambd` pre-rounded to bf16 RNE to match device |
| `ttnn.fill` | `fill_value` | exact (bit-for-bit), excluding sign-of-zero |
| `ttnn.leaky_relu` | `negative_slope` | ULP ≤ 1 |
| `ttnn.softshrink` | `lambd` (default 0.5) | ULP ≤ 1 |
| `ttnn.elu` | `alpha` (default 1.0, incl. 0 and negative) | ULP ≤ 1, excluding a narrow `exp(x)-1` cancellation band |
| `ttnn.celu` | `alpha` (default 1.0) | ULP ≤ 1, same cancellation band as elu, scaled by alpha |
| `ttnn.rpow` | `exponent` | PCC ≥ 0.99 (transcendental exp2/log2 SFPU chain), excluding an extreme-magnitude SFPU limitation |
| `ttnn.softcap` | `beta` | ULP ≤ 2 (post-FTZ) + PCC ≥ 0.9999; Blackhole only |

## What was added

- **`test_unary_category4_bfloat16.py`** — exhaustive normal bf16 sweep (32,512 positive
  + 32,512 negative normals per (op, param) pair) against each op's golden function,
  gated on exact-equality / ULP / PCC per the table above. Also exercises the preallocated
  `output_tensor=`/`queue_id=` call path exhaustively for `heaviside` and `leaky_relu`.
- Redundant small random-sample bf16 tests removed as superseded:
  - `test_activation.py`: `test_scalarB_elu`, `test_scalarB_hardshrink`,
    `test_scalarB_heaviside`, `test_scalarB_leaky_relu`, `test_scalarB_softshrink`
  - `test_fill.py`: `test_fill` (bfloat16 case)
  - `test_unary.py`: `test_relu_min`, `test_relu_max` (bfloat16-only exact-equality case)
  - `test_unary_ops_ttnn.py`: `test_unary_heaviside_ttnn`, `test_unary_leaky_relu_ttnn`
    (superseded once their `output_tensor=`/`queue_id=` path was folded into the
    exhaustive suite)

  fp32/int32/uint32/uint16/uint8/bfloat8_b/bfloat4_b dtype tests, edge-case tests
  (zero-beta guard, alpha=inf/nan), and tests explicitly cited as complementary
  (elu/celu cancellation-band `allclose` tests, `rpow`'s existing PCC coverage,
  `softcap`'s full-domain overshoot-bound tests) were left untouched.

  Note: `category4_ulp.py` (an exhaustive ULP/PCC characterization sweep used only to
  derive the thresholds above, mirroring `category3_ulp.py`) is a local analysis script
  and is **not** part of this PR / not CI-gating — it is not included in the diff.

## Accuracy notes / artifacts

| Op | Artifact handled explicitly |
|---|---|
| `heaviside` | `-0.0` excluded: device reads the sign bit and treats `-0.0` as negative (returns 0), while torch treats `+0.0`/`-0.0` as numerically equal (returns `value`) — confirmed on real hardware. |
| `fill` | `fill_value = -0.0` compared by value, not bits: device canonicalizes a `-0.0` fill_value to `+0.0` — confirmed on real hardware. All other `fill_value`s compared bit-exact; golden is built from the bf16-*truncated* fill_value (device truncates rather than round-to-nearest-even) so non-representable values (e.g. `0.1`, fp32 max) agree too. |
| `elu` / `celu` | `x ∈ [-0.28515625, ~1.166e-38]` excluded (scaled by `alpha` for celu, since it evaluates `exp(x/alpha)-1`): `exp(x)-1` loses >1 ULP to catastrophic cancellation in this narrow band; already characterized/covered via `allclose` in `test_elu.py` / `test_celu_21f.py`. |
| `leaky_relu` / `softshrink` | Subnormal *output* elements flushed to 0 on golden (device FTZ) before the ULP comparison; the mask uses the exact (non-bf16-rounded) result so a bf16-rounding artifact at the subnormal/normal boundary can't hide a real device mismatch, and `result` itself is left untouched so a device that fails to flush is still caught. |
| `rpow` | PCC/allclose computed only over elements where both golden and device outputs are finite *and* `|x * log2(exponent)| < 1e30`; beyond that magnitude the device's `exp2` argument reduction breaks down and returns a bogus finite value instead of overflowing — a real SFPU limitation, not a golden/test issue. Non-finite classification (NaN vs Inf, and Inf sign) is checked exactly within the reliable range. |
| `softcap` | Blackhole only; skipped elsewhere. `beta=25.0` intentionally omitted here (already covered exhaustively by `test_unary.py::test_softcap_bfloat16_full_domain`); the masked-out near-zero region is bounded, not just skipped. Hardware-verified on Blackhole: PCC ≈ 0.99999997, ULP=0 for ≥97.9% of in-domain elements. |

## Labels
op_cat: eltwise, MCW

### PR DESCRIPTION
**Pull request by abdulla-TT**
### PR Category
Test Only

### Summary
Closes #56823 (Category 4 sub-issue of #53613)

Adds exhaustive bfloat16 accuracy tests for the 11 ops with a float/scalar
parameter: `elu`, `heaviside`, `leaky_relu`, `relu_max`, `relu_min`, `rpow`,
`celu`, `softcap`, `fill`, `hardshrink`, `softshrink`.

- `test_unary_category4_bfloat16.py`: full bf16 sweep per op, gated by exact
  equality / ULP (≤1-2) / PCC depending on the op's precision profile. Also
  covers the `output_tensor=`/`queue_id=` path for `heaviside` & `leaky_relu`.
- Removed now-redundant random-sample bf16 tests superseded by the above, in
  `test_activation.py`, `test_fill.py`, `test_unary.py`, `test_unary_ops_ttnn.py`.

### Notes for reviewers
- `heaviside` excludes `-0.0`: device treats it as negative, torch doesn't.
- `leaky_relu`/`softshrink` flush subnormal outputs to 0 before ULP compare
  (device FTZ).
- `softcap` runs on Blackhole only.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:abdulla/unary_cat4_bf16)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:abdulla/unary_cat4_bf16)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+# ─────────────────────────────────────────────────────────────────────────────
+
+# Characterized (default alpha=1.0) and covered via allclose in test_elu.py /
+# test_celu_21f.py. elu's band is in x directly (alpha only scales the output
+# after cancellation); celu's band is in x/alpha (it evaluates exp(x/alpha)-1),
+# so the celu mask below scales the band by alpha.
+_ELU_CANCELLATION_BAND = (-0.28515625, 1.1663108012064884e-38)
+
+
+@pytest.mark.parametrize("alpha", [1.0, 0.5, 2.0, 0.1])
```
Comment (VirdhatchaniKN) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py]: The linked issue and PR body both promise elu coverage for `alpha` "incl. 0 and negative", and the removed `test_scalarB_elu` swept `[-0.5, 0, 0.5]`, but this sweep only uses positive alphas. The negative branch multiplies by `alpha`, so zero/negative-alpha handling is now untested — add non-positive alpha values here to match the stated criteria.

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    preallocated_result = ttnn.to_torch(preallocated_output)
+    assert torch.equal(preallocated_result, golden), (
+        f"heaviside(value={value}) via output_tensor=/queue_id= diverged for "
+        f"{int((preallocated_result != golden).sum().item())} of {preallocated_result.numel()} elements"
+    )
+
+
+@pytest.mark.parametrize("upper_limit", [0.0, 1.0, 3.0, 100.0, -1.0])
+def test_relu_max_op(device, upper_limit):
+    """output_i = min(max(x_i, 0), upper_limit). upper_limit=-1.0 covers the
```
Comment (VirdhatchaniKN) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py]: The relu_max golden is `relu(min(x, upper_limit))` = `max(0, min(x, upper_limit))`, so `upper_limit=-1.0` collapses to 0, not -1. This docstring's `min(max(x, 0), upper_limit)` is the reversed order and predicts -1 (same inversion in the header block near line 15).

**Review by VirdhatchaniKN — COMMENTED** [review]
Two findings inline (test-only, non-blocking).

**Review by VirdhatchaniKN — APPROVED** [review]
Re-review: both prior findings (elu non-positive alpha coverage, relu_max docstring) are addressed. LGTM.

**Review by mateusznowakTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+        not nan_mismatch.any()
+    ), f"rpow(exponent={exponent}) NaN classification diverged for {int(nan_mismatch.sum().item())} elements"
+    golden_inf, result_inf = torch.isinf(golden), torch.isinf(result)
+    sign_mismatch = golden_inf & result_inf & reliable & (torch.sign(golden) != torch.sign(result))
+    assert (
+        not sign_mismatch.any()
+    ), f"rpow(exponent={exponent}) infinity sign diverged for {int(sign_mismatch.sum().item())} elements"
+
+    finite = golden_finite & result_finite & reliable
+    assert_with_pcc(golden[finite], result[finite], pcc=0.99)
```
Comment (KalaivaniMCW) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py:384]: Pls check this out whether it is (golden , expected )

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    assert_allclose(result[finite], golden[finite], atol=1e-2, rtol=0.1)
```
Comment (KalaivaniMCW) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_category4_bfloat16.py:385]: Pls check this out whether it is (golden , expected )
